### PR TITLE
Introduce Get Menu Functionality

### PR DESCRIPTION
**Pull request by imRohan**
The Intelligent Foods Partner API includes a APi endpoint which returns the Menu along with its associated Menu Items. We'd like to incorporate that API call as a feature of this gem.

This change addresses the need by:
* Introducing a new #find method in IntelligentFoods::Menu
* Adding a new resource IntelligentFoods::MenuItem

### CODE REVIEW

File: lib/intelligent_foods/resources/menu.rb [hunk 1]
```diff
@@ -1,23 +1,43 @@
...
+      menu = build(data)
+      menu.items = menu_items
+      menu
     end
 
     def self.all
       uri = URI("#{IntelligentFoods.base_url}/menus")
       request = Net::HTTP::Get.new(uri)
       client = IntelligentFoods.client
+      puts client.inspect
```
Comment (maxgorbatenko) [inline lib/intelligent_foods/resources/menu.rb:21]: Seems to be left debug statements, unless I am misunderstanding those puts.

File: lib/intelligent_foods/api_client.rb [hunk 2]
```diff
@@ -4,26 +4,29 @@ module IntelligentFoods
   class ApiClient
     attr_reader :id, :secret, :access_token
 
-    def initialize(id: nil, secret: nil, client: nil)
+    def initialize(id: nil, secret: nil, client: nil, access_token: nil)
```
Comment (chriswoodford) [inline lib/intelligent_foods/api_client.rb:7]: this looks like an orthogonal change that should go in it's own PR? which should probably also include some refactoring to this class (and/or an additional class) to better support the change.

File: spec/fixtures/menu_response.json [hunk 3]
```diff
@@ -0,0 +1,102 @@
+{
```
Comment (chriswoodford) [inline spec/fixtures/menu_response.json:1]: should be spec/support/fixtures

**Review by chriswoodford — CHANGES_REQUESTED** [review]
split this into 2 PRs. let's get the authentication right and locked in. make the change easy, then make the easy change

**Review by maxgorbatenko — APPROVED** [review]
Looks good to me.

**Review by maxgorbatenko — APPROVED** [review]

File: spec/intelligent_foods/resources/menu_spec.rb [hunk 4]
```diff
@@ -24,4 +24,79 @@
...
   end
+
+  describe ".find" do
+    it "returns the menu" do
+      menu_id = "2023-01-01"
+      body = build_menu_response(menu_id: menu_id)
+      response = build_response(body: body)
+      stub_api_response response: response
+
+      result = IntelligentFoods::Menu.find(menu_id)
```
Comment (jmazur) [inline spec/intelligent_foods/resources/menu_spec.rb:35]: when working within the same module you dont need to use the namespace

**Review by jmazur — CHANGES_REQUESTED** [review]
Looking good, just the 1 style comment from me

File: spec/intelligent_foods/resources/menu_spec.rb [hunk 5]
```diff
@@ -24,4 +24,79 @@
...
+
+    it "returns an instance of IntelligentFoods::Menu" do
+      menu_id = "2023-01-01"
+      body = build_menu_response(menu_id: menu_id)
+      response = build_response(body: body)
+      stub_api_response response: response
+
+      result = IntelligentFoods::Menu.find(menu_id)
+
+      expect(result).to be_an_instance_of IntelligentFoods::Menu
```
Comment (chriswoodford) [inline spec/intelligent_foods/resources/menu_spec.rb:48]: prefer not to check for types. it's a dynamic language. consider checking for an attribute value or removing these tests

File: spec/intelligent_foods/resources/menu_spec.rb [hunk 6]
```diff
@@ -24,4 +24,79 @@
...
+      menu_items = stubbed_menu_items
+      expected_items_count = menu_items.count
+      body = build_menu_response(menu_id: menu_id, menu_items: menu_items)
+      response = build_response(body: body)
+      stub_api_response response: response
+      menu = IntelligentFoods::Menu.find(menu_id)
+
+      result = menu.items.count
+
+      expect(result).to eq expected_items_count
```
Comment (chriswoodford) [inline spec/intelligent_foods/resources/menu_spec.rb:75]:
```suggestion
      expect(result).to eq(expected_items_count)
```

File: spec/intelligent_foods/resources/menu_spec.rb [hunk 7]
```diff
@@ -24,4 +24,79 @@
...
+    it "assigns correct number of items" do
+      menu_id = "2023-01-01"
+      menu_items = stubbed_menu_items
+      expected_items_count = menu_items.count
+      body = build_menu_response(menu_id: menu_id, menu_items: menu_items)
+      response = build_response(body: body)
+      stub_api_response response: response
+      menu = IntelligentFoods::Menu.find(menu_id)
+
+      result = menu.items.count
```
Comment (chriswoodford) [inline spec/intelligent_foods/resources/menu_spec.rb:73]:
```suggestion
      result = menu.items.size
```

prefer `size` over `count`

File: lib/intelligent_foods/resources/object.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+module IntelligentFoods
+  class Object < OpenStruct
+    def self.build(data)
+      JSON.parse(data.to_json, object_class: self)
```
Comment (chriswoodford) [inline lib/intelligent_foods/resources/object.rb:6]: nice 👍

File: spec/support/helpers/api_helper.rb [hunk 9]
```diff
@@ -30,4 +32,20 @@ def build_response(body: {}, http_status_code: 200)
   def build_encoded_token(id:, secret:)
     Base64.strict_encode64("#{id}:#{secret}")
   end
+
+  def read_stub_file
```
Comment (chriswoodford) [inline spec/support/helpers/api_helper.rb:36]: this is a bit vague... what is a "stub file"? probably should be "fixture". the menu api response files is hardcoded in the method.. either pass that in as a method param or consider calling this `read_menu_api_response`

File: spec/intelligent_foods/resources/menu_spec.rb (hunk 6 above)
Comment (chriswoodford) [reply spec/intelligent_foods/resources/menu_spec.rb:75]: prefer using the parens for test expectations

**Review by chriswoodford — CHANGES_REQUESTED** [review]
with the next PR, it's time to introduce a fake to consolidate the API testing code

File: spec/support/helpers/api_helper.rb (hunk 9 above)
Comment (maxgorbatenko) [reply spec/support/helpers/api_helper.rb]: renamed the method.

File: spec/intelligent_foods/resources/menu_spec.rb [hunk 10]
```diff
@@ -24,4 +24,55 @@
...
+
+      result = IntelligentFoods::Menu.find(menu_id)
+
+      expect(result.id).to eq(menu_id)
+    end
+
+    it "assigns correct number of items" do
+      menu_id = "2023-01-01"
+      menu_items = stubbed_menu_items
+      expected_items_count = menu_items.size
```
Comment (chriswoodford) [inline spec/intelligent_foods/resources/menu_spec.rb:43]: this is a little too mystery guest for my taste. can you tell the stubbed menu items how many items you want and then hard code the expected items count amount

**Review by chriswoodford — CHANGES_REQUESTED** [review]
just one comment on the one of the specs

File: spec/intelligent_foods/resources/menu_spec.rb (hunk 10 above)
Comment (chriswoodford) [reply spec/intelligent_foods/resources/menu_spec.rb:43]: https://thoughtbot.com/blog/mystery-guest
https://thoughtbot.com/upcase/decks/5/flashcards/27?mode=reviewing

**Review by chriswoodford — CHANGES_REQUESTED** [review]

**Review by chriswoodford — APPROVED** [review]

File: spec/intelligent_foods/resources/menu_spec.rb (hunk 4 above)
Comment (maxgorbatenko) [reply spec/intelligent_foods/resources/menu_spec.rb:35]: @jmazur Very possible that I am misunderstanding something, but deleting that namespace breaks the tests. Possibly not doing something right.
